### PR TITLE
GH-39976: [C++] Fix out-of-line data size calculation in BinaryViewBuilder::AppendArraySlice

### LINKED ISSUE #39976: [C++] ASAN reports heap buffer overflow in `TestArray.TestAppendArraySlice`
### Describe the bug, including details regarding any error messages, version, and platform.

This is a failure that should have been reported by our ASAN check but is somehow prevented by the problem discussed in #39973 . There is a a living demo in https://github.com/zanmato1984/arrow/actions/runs/7752895694/job/21143188221?pr=3#step:6:3370. One can also reproduce this issue in local by:
```
cmake --preset ninja-debug -DARROW_USE_ASAN=ON -DARROW_JEMALLOC=OFF -DARROW_MIMALLOC=OFF ..
ninja -j8
./debug/arrow-array-test
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

Fix the bug in `BinaryViewBuilder::AppendArraySlice` that, when calculating out-of-line data size, the array is wrongly iterated.

### What changes are included in this PR?

Fix and UT.

### Are these changes tested?

UT included.

### Are there any user-facing changes?

No.

* Closes: #39976

### CODE REVIEW

**Comment by zanmato1984** [comment]
cc @pitrou @bkietz

**Review by mapleFU — APPROVED** [review]
+1

File: cpp/src/arrow/array/array_test.cc [hunk 1]
```diff
@@ -905,6 +905,25 @@ TEST_F(TestArray, TestAppendArraySlice) {
...
+  ASSERT_OK_AND_ASSIGN(auto src, src_builder.Finish());
+  ASSERT_OK(src->ValidateFull());
+
+  ArraySpan span;
+  span.SetMembers(*src->data());
+  BinaryViewBuilder dst_builder(pool_);
+  ASSERT_OK(dst_builder.AppendArraySlice(span, 1, 1));
+  ASSERT_EQ(1, dst_builder.length());
+  ASSERT_OK_AND_ASSIGN(auto dst, dst_builder.Finish());
+  ASSERT_OK(dst->ValidateFull());
```
Comment (pitrou) [inline cpp/src/arrow/array/array_test.cc:924]: Can you also test the array value using `AssertArraysEqual`?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a lot for fixing this. Just one suggestion below.

File: cpp/src/arrow/array/array_test.cc (hunk 1 above)
Comment (zanmato1984) [reply cpp/src/arrow/array/array_test.cc:926]: Addressed.

**Review by pitrou — APPROVED** [review]

**Pull request merged**